### ISSUE TITLE
[FreeBSD] zrepl and zfs send sometimes get into a deadlock-like state

### ISSUE THREAD

**Issue by avg-I** [zrepl/zrepl#495]
I periodically observe a problem where a zrepl instance configured with a single push job stops doing its work.
In that state it logs messages like
```
zrepl[12895]: job=backup subsystem=snapshot span=Uv38$DEgt$F9Bq$F9Bq.z4Xn msg="callback channel is full, discarding snapshot update event"
```

I see that the zrepl process is waiting on a zfs send process.
```
   0 12895 12894  3  25  0  751192  19648 uwait    I     -      1:41.38 /usr/local/bin/zrepl --config /usr/local/etc/zrepl/zrepl.yml daemon
   0 13287 12895  1  20  0   17452   5736 tq_drain D     -      0:00.50 zfs send -t 1-f2979c6bf-d8-789c636064000310a500c4ec50360710e72765a5269730307c81aac1904f4b2b4e2d01c958b1c1e4d990e4932a4b528b81b484dd7e366cfa4bf2d34b33531818f4f3ea56ab54f87ff34092e704cbe725e6a632301415e4e7e7e8971617e967e4e7a63a5415a516e4e81a1918191a98191ac61b589a5a1a99c41b1818302001007cb71f86
```
The send process is stuck waiting for a ZFS helper thread to complete writing to a pipe.
``` 
PID    TID COMM                TDNAME              KSTACK                       
13287 100583 zfs                 -                   mi_switch+0x178 sleepq_switch+0xfc _sleep+0x27c taskqueue_drain+0x100 taskq_wait_id+0x2c spa_taskq_dispatch_sync+0xa8 dump_bytes+0x30 dump_record+0x14c dmu_dump_write+0x5d8 dmu_send_impl+0xd70 dmu_send+0x5b8 zfs_ioc_send_new+0x164 zfsdev_ioctl_common+0x534 zfsdev_ioctl+0x150 devfs_ioctl+0xfc vn_ioctl+0x108 devfs_ioctl_f+0x20 kern_ioctl+0x2ec 
13287 100807 zfs                 send_traverse_threa mi_switch+0x178 sleepq_switch+0xfc sleepq_catch_signals+0x2d4 sleepq_wait_sig+0x8 _cv_wait_sig+0x164 bqueue_enqueue_impl+0x78 send_cb+0x530 traverse_visitbp+0x1c8 traverse_visitbp+0x6ec traverse_dnode+0xb4 traverse_visitbp+0x8c0 traverse_visitbp+0x6ec traverse_visitbp+0x6ec traverse_visitbp+0x6ec traverse_visitbp+0x6ec traverse_visitbp+0x6ec traverse_dnode+0xb4 traverse_visitbp+0x520 
13287 100809 zfs                 send_merge_thread   mi_switch+0x178 sleepq_switch+0xfc sleepq_catch_signals+0x2d4 sleepq_wait_sig+0x8 _cv_wait_sig+0x164 bqueue_enqueue_impl+0x78 send_merge_thread+0x12c fork_exit+0x74 fork_trampoline+0x14 
13287 100810 zfs                 send_reader_thread  mi_switch+0x178 sleepq_switch+0xfc sleepq_catch_signals+0x2d4 sleepq_wait_sig+0x8 _cv_wait_sig+0x164 bqueue_enqueue_impl+0x78 send_reader_thread+0x14c fork_exit+0x74 fork_trampoline+0x14
```
```
    0 100394 kernel              z_fr_iss_1          mi_switch+0x178 sleepq_switch+0xfc sleepq_catch_signals+0x2d4 sleepq_wait_sig+0x8 _sleep+0x270 pipe_write+0x8cc zfs_file_write_impl+0xac zfs_file_write+0x24 dump_bytes_cb+0x20 taskq_run+0x1c taskqueue_run_locked+0x188 taskqueue_thread_loop+0x9c fork_exit+0x74 fork_trampoline+0x14
```
The other end of the pipe is apparently in the zrepl process.
But it seems that the zrepl process is not reading from the pipe.
At least I cannot identify a thread that would be doing that.
```
12895 100608 zrepl               -                   mi_switch+0x178 sleepq_switch+0xfc sleepq_catch_signals+0x2d4 sleepq_wait_sig+0x8 _sleep+0x270 umtxq_sleep+0x1ac do_wait+0x228 __umtx_op_wait_uint_private+0x60 sys__umtx_op+0x7c do_el0_sync+0x4a4 handle_el0_sync+0x90 
12895 100756 zrepl               -                   mi_switch+0x178 sleepq_switch+0xfc sleepq_catch_signals+0x2d4 sleepq_timedwait_sig+0x14 _sleep+0x248 umtxq_sleep+0x1ac do_wait+0x228 __umtx_op_wait_uint_private+0x60 sys__umtx_op+0x7c do_el0_sync+0x4a4 handle_el0_sync+0x90 
12895 100757 zrepl               -                   mi_switch+0x178 sleepq_switch+0xfc sleepq_catch_signals+0x2d4 sleepq_wait_sig+0x8 _sleep+0x270 umtxq_sleep+0x1ac do_wait+0x228 __umtx_op_wait_uint_private+0x60 sys__umtx_op+0x7c do_el0_sync+0x4a4 handle_el0_sync+0x90 
12895 100758 zrepl               -                   mi_switch+0x178 sleepq_switch+0xfc sleepq_catch_signals+0x2d4 sleepq_wait_sig+0x8 _sleep+0x270 umtxq_sleep+0x1ac do_wait+0x228 __umtx_op_wait_uint_private+0x60 sys__umtx_op+0x7c do_el0_sync+0x4a4 handle_el0_sync+0x90 
12895 100759 zrepl               -                   mi_switch+0x178 sleepq_switch+0xfc sleepq_catch_signals+0x2d4 sleepq_timedwait_sig+0x14 _sleep+0x248 kqueue_kevent+0x9c8 kern_kevent_fp+0xb0 kern_kevent+0x88 kern_kevent_generic+0x78 freebsd11_kevent+0x68 do_el0_sync+0x4a4 handle_el0_sync+0x90 
12895 100760 zrepl               -                   mi_switch+0x178 sleepq_switch+0xfc sleepq_catch_signals+0x2d4 sleepq_wait_sig+0x8 _sleep+0x270 umtxq_sleep+0x1ac do_wait+0x228 __umtx_op_wait_uint_private+0x60 sys__umtx_op+0x7c do_el0_sync+0x4a4 handle_el0_sync+0x90 
12895 100761 zrepl               -                   mi_switch+0x178 sleepq_switch+0xfc sleepq_catch_signals+0x2d4 sleepq_wait_sig+0x8 _sleep+0x270 umtxq_sleep+0x1ac do_wait+0x228 __umtx_op_wait_uint_private+0x60 sys__umtx_op+0x7c do_el0_sync+0x4a4 handle_el0_sync+0x90 
12895 100762 zrepl               -                   mi_switch+0x178 sleepq_switch+0xfc sleepq_catch_signals+0x2d4 sleepq_wait_sig+0x8 _sleep+0x270 umtxq_sleep+0x1ac do_wait+0x228 __umtx_op_wait_uint_private+0x60 sys__umtx_op+0x7c do_el0_sync+0x4a4 handle_el0_sync+0x90 
12895 100763 zrepl               -                   mi_switch+0x178 sleepq_switch+0xfc sleepq_catch_signals+0x2d4 sleepq_wait_sig+0x8 _sleep+0x270 umtxq_sleep+0x1ac do_wait+0x228 __umtx_op_wait_uint_private+0x60 sys__umtx_op+0x7c do_el0_sync+0x4a4 handle_el0_sync+0x90 
12895 100764 zrepl               -                   mi_switch+0x178 sleepq_switch+0xfc sleepq_catch_signals+0x2d4 sleepq_wait_sig+0x8 _sleep+0x270 umtxq_sleep+0x1ac do_wait+0x228 __umtx_op_wait_uint_private+0x60 sys__umtx_op+0x7c do_el0_sync+0x4a4 handle_el0_sync+0x90 
12895 100768 zrepl               -                   mi_switch+0x178 sleepq_switch+0xfc sleepq_catch_signals+0x2d4 sleepq_wait_sig+0x8 _sleep+0x270 umtxq_sleep+0x1ac do_wait+0x228 __umtx_op_wait_uint_private+0x60 sys__umtx_op+0x7c do_el0_sync+0x4a4 handle_el0_sync+0x90 
12895 100769 zrepl               -                   mi_switch+0x178 sleepq_switch+0xfc sleepq_catch_signals+0x2d4 sleepq_wait_sig+0x8 _sleep+0x270 kern_wait6+0x3d8 sys_wait6+0x68 do_el0_sync+0x4a4 handle_el0_sync+0x90 
12895 100777 zrepl               -                   mi_switch+0x178 sleepq_switch+0xfc sleepq_catch_signals+0x2d4 sleepq_wait_sig+0x8 _sleep+0x270 umtxq_sleep+0x1ac do_wait+0x228 __umtx_op_wait_uint_private+0x60 sys__umtx_op+0x7c do_el0_sync+0x4a4 handle_el0_sync+0x90
```

At the moment I am not where the problem might be: zrepl, go run-time or FreeBSD.
My current understanding based on an ongoing kgdb session is that the pipe was full at some moment, then zrepl read some but not all data from it. After that neither zrepl returned to read more data nor the writer was notified that it can write more data.

**Comment by problame** [comment]
This sounds like a duplicate of https://github.com/zrepl/zrepl/issues/411

Please try pinning the zrepl process to a single core and see whether the issue still occurs.

**Comment by avg-I** [comment]
I saw #411 but to me it did not look completely similar as that issue does not mention local pipes.
Anyway, I will try the suggestion and report back after a while.

**Comment by avg-I** [comment]
I think that my problem is caused by a flaky network (backup over relatively weak WiFi).
It looks that there are two TCP connection between a client system (doing the push) and a server.
As far as I can tell, the data is sent over one of the connections, the purpose of the other is not clear to me.
From what I see, during a flaky network event the data connection can be lost. The other connection can be lost as well but it gets re-established. In any case, after that point the client zrepl stops sending any data and ends up in the state I described initially.

Some data.
This is what I see during the normal network on the client side:
```
root     zrepl      23346 9  tcp4   192.168.0.35:32789    192.168.0.88:7888
root     zrepl      23346 10 tcp4   192.168.0.35:20934    192.168.0.88:7888
```
But after the flaky network it would be something like:
```
root     zrepl      18927 9  tcp4   192.168.0.35:46630    192.168.0.88:7888
```
Using tcpdump I can see that the client zrepl sends only very short packets over the remaining connection.

Also, I should add that I am running zrepl with the default log level and I do not see anything logged after the data connection is lost.

**Comment by problame** [comment]
> I think that my problem is caused by a flaky network (backup over relatively weak WiFi).
It looks that there are two TCP connection between a client system (doing the push) and a server.
As far as I can tell, the data is sent over one of the connections, the purpose of the other is not clear to me.

One is rather long-lived (gRPC-based control protocol), the other connection(s) are for ZFS send stream data transfer.

> From what I see, during a flaky network event the data connection can be lost. The other connection can be lost as well but it gets re-established. In any case, after that point the client zrepl stops sending any data and ends up in the state I described initially.

Yes, gRPC has a re-dial mechanism, but the data conns die and, in theory, zrepl should eventually retry the replication, thereby eventually re-creating the data connections. Default retry timeout is 10min.

> Using tcpdump I can see that the client zrepl sends only very short packets over the remaining connection.

That ought to be heartbeats.

Could you have a look at `zrepl status` on the active side? What does it report in the `Replication` section?

**Comment by avg-I** [comment]
Unfortunately (fortunately?) I cannot reproduce the issue any more.
I managed to improve the network reliability and the sync started working okay.
I intentionally degraded the network again, but the issue does not come back.
Perhaps, it also requires full snapshots or snapshots above certain size.
When I had the problem originally the problematic system could not get through the initial backup.

**Comment by avg-I** [comment]
It finally happened.
First, the processes:
```
   0  1106  1105 1  26   0 751564 28924 uwait    I     -      7:49.26 /usr/local/bin/zrepl --config /usr/local/etc/zrepl/zrepl.yml daemon
   0 34010  1106 0  20   0  17452  6740 tq_drain D     -      0:00.07 zfs send -i rpool/var/log@zrepl-20210804_062954_000 rpool/var/log@zrepl-20210804_142954_000
```

Now, zrepl status:
```
Job: backup                                                                                                                                             
Type: push                                                                                                                                              
                                                                                                                                                        
Replication:                                                                                                                                            
   Attempt #1                                                                                                                                           
   Status: fan-out-filesystems                                                                                                                          
   Started: 2021-08-04 14:29:55 +0000 UTC (lasting 15h40m38s)                                                                                           
   Progress: [>--------------------------------------------------] 1.8 KiB / 751.6 KiB @ 0 B/s                                                          
     rpool/ROOT                         DONE (step 1/1, 624 B/624 B)                                                                                    
     rpool/ROOT/2021-02-01-4bbfe4bf08d6 DONE (step 1/1, 624 B/624 B)                                                                                    
     rpool/ROOT/2021-06-08              STEPPING (step 0/1, 0 B/157.8 KiB) next: @zrepl-20210804_062954_000 => @zrepl-20210804_142954_000 (encrypted=no)
     rpool/usr/home                     STEPPING (step 0/1, 0 B/77.1 KiB) next: @zrepl-20210804_062954_000 => @zrepl-20210804_142954_000 (encrypted=no) 
     rpool/var/log                      STEPPING (step 0/1, 0 B/514.8 KiB) next: @zrepl-20210804_062954_000 => @zrepl-20210804_142954_000 (encrypted=no)
     rpool/var/mail                     DONE (step 1/1, 624 B/624 B)                                                                                    
                                                                                                                                                        
Pruning Sender:                                                                                                                                         
   ...                                                                                                                                                  
                                                                                                                                                        
Pruning Receiver:                                                                                                                                       
   ...                                                                                                                                                  
                                                                                                                                                        
Snapshotting:                                                                                                                                           
   Status: Waiting                                                                                                                                      
   Sleep until: 2021-08-05 06:29:52.826349479 +0000 UTC                                                                                                 
   rpool/ROOT                         SnapDone 300ms snap name: "zrepl-20210804_222954_000"                                                             
   rpool/ROOT/2021-02-01-4bbfe4bf08d6 SnapDone 300ms snap name: "zrepl-20210804_222954_000"                                                             
   rpool/ROOT/2021-06-08              SnapDone 300ms snap name: "zrepl-20210804_222953_000"                                                             
   rpool/usr/home                     SnapDone 400ms snap name: "zrepl-20210804_222953_000"                                                             
   rpool/var/log                      SnapDone 200ms snap name: "zrepl-20210804_222953_000"                                                             
   rpool/var/mail                     SnapDone 300ms snap name: "zrepl-20210804_222954_000"                                                             
```

And there is only one tcp connection from zrepl to the backup host:
```
$ sockstat | fgrep zrepl
root     zrepl      1106  6  dgram  -> /var/run/log
root     zrepl      1106  7  stream /var/run/zrepl/control
root     zrepl      1106  9  tcp4   192.168.0.136:55939   192.168.0.88:7888
```

**Comment by barrylustig** [comment]
I run into this problem fairly regularly (every 3-7 days).  I see the the same symptoms as Andriy: zrepl status shows an ever increasing "lasting" time in Replication and there is no progress happening.

**Comment by problame** [comment]
@barrylustig also on FreeBSD?

**Comment by problame** [comment]
> And there is only one tcp connection from zrepl to the backup host:

@avg-I it would be really useful to know whether that TCP connection is the gRPC connection (long-lived) or the one-off TCP connection used for the data transfer (`zfs send -i`). Any way to identify that? Both protocols use heartbeats, so tcpdump should work.

Also, do you know what that `tq_drain` state is (`zfs send -i` ps output)?

**Comment by barrylustig** [comment]
> @barrylustig also on FreeBSD?

Yes. 13.0-RELEASE-p3

**Comment by problame** [comment]
Could you try pinning the zrepl daemon process to a single logical CPU using cpuset? See https://github.com/zrepl/zrepl/issues/411#issuecomment-762160424

**Comment by barrylustig** [comment]
> Could you try pinning the zrepl daemon process to a single logical CPU using cpuset? See https://github.com/zrepl/zrepl/issues/411#issuecomment-762160424

I just locked it down to a single core. It usually takes at least a few days to trigger, so I'll report back later in the week unless it fails.

**Comment by avg-I** [comment]
> @avg-I it would be really useful to know whether that TCP connection is the gRPC connection (long-lived) or the one-off TCP connection used for the data transfer (`zfs send -i`). Any way to identify that? Both protocols use heartbeats, so tcpdump should work.

Based on my earlier observations, when I saw two connections initially and then only one remaining, the remaining connection is the control connection.

> Also, do you know what that `tq_drain` state is (`zfs send -i` ps output)?

Yes. This is available in the very first comment: dump_bytes -> spa_taskq_dispatch_sync -> taskq_wait_id -> taskqueue_drain.
As I described originally, this is stuck because the pipe from zfs send to zrepl is full and zrepl is not reading from it.
And I guess that that's because zrepl lost the data connection without being properly aware of it, so to say.
So, my impression is that zrepl is no longer "pumping" the data, but it is not doing anything to recover / clean up / etc either.

P.S.
And I already tried pinning the process but it did not help in my case.
My gut feeling is that the problem is about handling some edge case (like a FreeBSD specific error code or signal) when a specific network condition happens.

P.P.S.
Links to the potentially useful information I provided earlier:
https://github.com/zrepl/zrepl/issues/495#issue-922687420
https://github.com/zrepl/zrepl/issues/495#issuecomment-863299946

**Comment by barrylustig** [comment]
> > Could you try pinning the zrepl daemon process to a single logical CPU using cpuset? See [#411 (comment)](https://github.com/zrepl/zrepl/issues/411#issuecomment-762160424)
> 
> I just locked it down to a single core. It usually takes at least a few days to trigger, so I'll report back later in the week unless it fails.

It failed about 10 hours ago.  Is there any information I can try and gather for you?  I see the same connection issue as Andriy:

```
[root@mail /var/log/rspamd]# sockstat | fgrep zrepl
root     zrepl      81788 6  dgram  -> /var/run/log
root     zrepl      81788 7  stream /var/run/zrepl/control
root     zrepl      81788 8  tcp4   192.168.222.1:43535   192.168.222.2:8888
```
and 
```
procstat -k 19841 (zfs send process)
  PID    TID COMM                TDNAME              KSTACK
19841 100844 zfs                 -                   mi_switch _sleep taskqueue_drain taskq_wait_id spa_taskq_dispatch_sync dump_bytes dump_record dmu_dump_write do_dump dmu_send_impl dmu_send zfs_ioc_send_new zfsdev_ioctl_common zfsdev_ioctl devfs_ioctl vn_ioctl devfs_ioctl_f kern_ioctl
19841 114951 zfs                 send_reader_thread  mi_switch sleepq_catch_signals sleepq_wait_sig _cv_wait_sig bqueue_enqueue_impl send_reader_thread fork_exit fork_trampoline
```
```
procstat -k 81788 (zrepl process)
  PID    TID COMM                TDNAME              KSTACK
81788 100673 zrepl               -                   mi_switch sleepq_catch_signals sleepq_wait_sig _sleep umtxq_sleep do_wait __umtx_op_wait_uint_private sys__umtx_op amd64_syscall fast_syscall_common
81788 100705 zrepl               -                   mi_switch sleepq_catch_signals sleepq_timedwait_sig _sleep umtxq_sleep do_wait __umtx_op_wait_uint_private sys__umtx_op amd64_syscall fast_syscall_common
81788 100706 zrepl               -                   mi_switch sleepq_catch_signals sleepq_wait_sig _sleep umtxq_sleep do_wait __umtx_op_wait_uint_private sys__umtx_op amd64_syscall fast_syscall_common
81788 100707 zrepl               -                   mi_switch sleepq_catch_signals sleepq_wait_sig _sleep kern_wait6 sys_wait6 amd64_syscall fast_syscall_common
81788 100708 zrepl               -                   mi_switch sleepq_catch_signals sleepq_wait_sig _sleep umtxq_sleep do_wait __umtx_op_wait_uint_private sys__umtx_op amd64_syscall fast_syscall_common
81788 100709 zrepl               -                   mi_switch sleepq_catch_signals sleepq_wait_sig _sleep umtxq_sleep do_wait __umtx_op_wait_uint_private sys__umtx_op amd64_syscall fast_syscall_common
81788 100710 zrepl               -                   mi_switch sleepq_catch_signals sleepq_wait_sig _sleep umtxq_sleep do_wait __umtx_op_wait_uint_private sys__umtx_op amd64_syscall fast_syscall_common
81788 100712 zrepl               -                   mi_switch sleepq_catch_signals sleepq_wait_sig _sleep umtxq_sleep do_wait __umtx_op_wait_uint_private sys__umtx_op amd64_syscall fast_syscall_common
81788 100713 zrepl               -                   mi_switch sleepq_catch_signals sleepq_wait_sig _sleep umtxq_sleep do_wait __umtx_op_wait_uint_private sys__umtx_op amd64_syscall fast_syscall_common
81788 100727 zrepl               -                   mi_switch sleepq_catch_signals sleepq_wait_sig _sleep umtxq_sleep do_wait __umtx_op_wait_uint_private sys__umtx_op amd64_syscall fast_syscall_common
81788 100730 zrepl               -                   mi_switch sleepq_catch_signals sleepq_wait_sig _sleep umtxq_sleep do_wait __umtx_op_wait_uint_private sys__umtx_op amd64_syscall fast_syscall_common
81788 100731 zrepl               -                   mi_switch sleepq_catch_signals sleepq_wait_sig _sleep umtxq_sleep do_wait __umtx_op_wait_uint_private sys__umtx_op amd64_syscall fast_syscall_common
81788 100751 zrepl               -                   mi_switch sleepq_catch_signals sleepq_timedwait_sig _sleep kqueue_kevent kern_kevent kern_kevent_generic freebsd11_kevent amd64_syscall fast_syscall_common
81788 101189 zrepl               -                   mi_switch sleepq_catch_signals sleepq_wait_sig _sleep umtxq_sleep do_wait __umtx_op_wait_uint_private sys__umtx_op amd64_syscall fast_syscall_common
```
The system logs where the failure happened:
```
Aug 10 03:10:12 mail zrepl[81788]: job=prod subsystem=repl span=Uv38$ImUo$u3FK$u3FK.YciE.7AX9 msg="receive request failed (might also be error on sender)" filesystem=zroot/ROOT/default rr="Filesystem:\"zroot/ROOT/default\" To:{Name:\"zrepl_20210810_100807_000\" Guid:14712106199920928990 CreateTXG:28695106 Creation:\"2021-08-10T03:08:08-07:00\"} ClearResumeToken:true ReplicationConfig:{protection:{Initial:GuaranteeIncrementalReplication Incremental:GuaranteeIncrementalReplication}}" err="stream: read error: read tcp 192.168.222.1:30306->192.168.222.2:8888: i/o timeout" errType=*stream.ReadStreamError
Aug 10 03:10:12 mail zrepl[79884]: 2021-08-10T03:10:12-07:00 [ERRO][prod][repl][Uv38$ImUo$u3FK$u3FK.YciE.7AX9]: receive request failed (might also be error on sender) errType="*stream.ReadStreamError" filesystem="zroot/ROOT/default" rr="Filesystem:\"zroot/ROOT/default\" To:{Name:\"zrepl_20210810_100807_000\" Guid:14712106199920928990 CreateTXG:28695106 Creation:\"2021-08-10T03:08:08-07:00\"} ClearResumeToken:true ReplicationConfig:{protection:{Initial:GuaranteeIncrementalReplication Incremental:GuaranteeIncrementalReplication}}" err="stream: read error: read tcp 192.168.222.1:30306->192.168.222.2:8888: i/o timeout"
Aug 10 03:38:29 mail zrepl[81788]: job=prod subsystem=snapshot span=Uv38$ImUo$N5J4$N5J4.dwMm msg="callback channel is full, discarding snapshot update event"
```

The data connection runs over a wireguard-encrypted tunnel to the remote end.  AFAICT, there were no glitches in the tunnel -- at least nothing was logged.

**Comment by problame** [comment]
Sorry for the delayed reply.

@avg-I 

> My gut feeling is that the problem is about handling some edge case (like a FreeBSD specific error code or signal) when a specific network condition happens.

Might very well be the case. But that's all abstracted away by Go / it's runtime. Which is known to have subtle bugs on FreeBSD....

@barrylustig 

The error in the syslog suggests that the TCP connection timed out.
Do I understand correctly that the `repl` job is a job in `pull` mode?

**Comment by barrylustig** [comment]
> Sorry for the delayed reply.
> 
> 
> 
> @avg-I 
> 
> 
> 
> > My gut feeling is that the problem is about handling some edge case (like a FreeBSD specific error code or signal) when a specific network condition happens.
> 
> 
> 
> Might very well be the case. But that's all abstracted away by Go / it's runtime. Which is known to have subtle bugs on FreeBSD....
> 
> 
> 
> @barrylustig 
> 
> 
> 
> The error in the syslog suggests that the TCP connection timed out.
> 
> Do I understand correctly that the `repl` job is a job in `pull` mode?

It's a `push` job:
```
jobs:
- name: prod
  type: push
  connect:
    type: tcp
    address: "192.168.222.2:8888"
  filesystems: {
    "<": true,
    "zroot/tmp": false,
    "zroot/docker<": false,
    "zroot/usr/ports<": false,
    "zroot/usr/src<": false,
    "zroot/var/crash<": false,
    "zroot/var/tmp<": false,
  }
#  send:
#    compressed: true
  replication:
    protection:
      initial: guarantee_incremental
      incremental: guarantee_incremental
...
```

**Comment by problame** [comment]
Curious. I think the following error message is emitted from the subsequent location in the source:

```
Aug 10 03:10:12 mail zrepl[79884]: 2021-08-10T03:10:12-07:00 [ERRO][prod][repl][Uv38$ImUo$u3FK$u3FK.YciE.7AX9]: receive request failed (might also be error on sender) errType="*stream.ReadStreamError" filesystem="zroot/ROOT/default" rr="Filesystem:\"zroot/ROOT/default\" To:{Name:\"zrepl_20210810_100807_000\" Guid:14712106199920928990 CreateTXG:28695106 Creation:\"2021-08-10T03:08:08-07:00\"} ClearResumeToken:true ReplicationConfig:{protection:{Initial:GuaranteeIncrementalReplication Incremental:GuaranteeIncrementalReplication}}" err="stream: read error: read tcp 192.168.222.1:30306->192.168.222.2:8888: i/o timeout"
```

https://github.com/zrepl/zrepl/blob/10a14a8c5016e8f7d3f145fdfd8ee76af274ede8/rpc/dataconn/stream/stream.go#L248-L250

(side note: it might be useful to have a toggle that dumps the stack trace recorded by `github.com/pkg/errors` )

Anyways, since it's a TCP read that is timing out here, it's most likely a wait for a reply message.
Still don't have an idea what's going on, but it seemed worth writing this insight down somewhere.

**Comment by avg-I** [comment]
Just want to add that today the issue happened again and this time I also got the same message "receive request failed (might also be error on sender)". Just in case, the message is on the push job side.
According to ps, zfs send was started at Fri 20 Aug 06:30:46 2021.
The message was at 06:30:56, 10 seconds later.
```
zrepl[1106]: job=backup subsystem=repl span=Uv38$/b/t$VSZo$VSZo.N3mm.CbQi msg="receive request failed (might also be error on sender)" err="dial tcp: i/o timeout" rr="Filesystem:\"rpool/var/log\"  To:{Name:\"zrepl-20210820_062955_000\"  Guid:101691263862493717  CreateTXG:1240327  Creation:\"2021-08-20T06:29:55Z\"}  ClearResumeToken:true  ReplicationConfig:{protection:{Initial:GuaranteeResumability  Incremental:GuaranteeResumability}}" errType=*net.OpError filesystem=rpool/var/log
```

The last log messages (about successful operations) on the sink time are from before 06:30:46, to be precise the very last message is timestamped 06:30:43.

**Comment by avg-I** [comment]
I went through the code and I have a guess.
If I understand the code correctly, SendStream::killAndWait in zfs.go first sends a signal, then waits for the command to exit and then closes the process's pipe. I think that that does not work on FreeBSD at least.
To recap, here is the state of zfs send:
```
  PID    TID COMM                TDNAME              KSTACK                       
11210 100541 zfs                 -                   mi_switch+0x178 sleepq_switch+0xfc _sleep+0x27c taskqueue_drain+0x100 taskq_wait_id+0x2c spa_taskq_dispatch_sync+0xa8 dump_bytes+0x30 dump_record+0x14c dmu_dump_write+0x5d8 dmu_send_impl+0xd70 dmu_send+0x5b8 zfs_ioc_send_new+0x164 zfsdev_ioctl_common+0x534 zfsdev_ioctl+0x150 devfs_ioctl+0xfc vn_ioctl+0x108 devfs_ioctl_f+0x20 kern_ioctl+0x2ec
```
taskqueue_drain is not interruptible, it calls msleep without PCATCH flag.
So, it is not possible to kill such process with any signal.
It must allowed to return from the kernel (or at least block in some other place where signals are checked) before it can be killed.
Demo:
```
# ps axwwl -p 11210
UID   PID PPID C PRI NI   VSZ  RSS MWCHAN   STAT TT     TIME COMMAND
  0 11210 1106 2  20  0 17452 6620 tq_drain D     -  0:00.07 zfs send -i rpool/var/log@zrepl-20210819_222956_000 rpool/var/log@zrepl-20210820_062955_000

# kill -9 11210

# ps axwwl -p 11210
UID   PID PPID C PRI NI   VSZ  RSS MWCHAN   STAT TT     TIME COMMAND
  0 11210 1106 2  20  0 17452 6620 tq_drain D     -  0:00.07 zfs send -i rpool/var/log@zrepl-20210819_222956_000 rpool/var/log@zrepl-20210820_062955_000
```

So, I think that at least on FreeBSD we should first close the pipe and, thus, unblock the process.
And then kill it if it does not exit by itself.

For example, killing the zrepl itself closes the pipe and the zfs send immediately terminates as well.

Finally, procstat of the zrepl shows that one of its threads is indeed in wait6 system call:
```
 1106 100569 zrepl               -                   mi_switch+0x178 sleepq_switch+0xfc sleepq_catch_signals+0x2d4 sleepq_wait_sig+0x8 _sleep+0x270 kern_wait6+0x3d8 sys_wait6+0x68
```

**Comment by problame** [comment]
@avg-I thanks for this excellent debugging work.

The way I see it: you are apparently right about the 'unkillable' behavior of `zfs send` that you described above.
My gut feeling is that this should be a FreeBSD bug, but regardless, we can work around it in zrepl.

I posted a patch, it should appear above. Could you try it out?

**Comment by avg-I** [comment]
I will test the patch after the weekend. Thank you!

P.S.
I would not call the FreeBSD behavior a bug, a peculiarity maybe.
And personally to me closing the pipe before killing the process makes more sense as the former seems a more graceful way to communicate with the process.
Without having looked at the patch yet, I'd say that closing the pipe and sending a signal (in either order) and only then waiting for the process to exit should be the most robust.

**Comment by problame** [comment]
The patch
* waits 200ms / SIGKILLs depending on whether it's an io.EOF error or not
* closes the pipe
* waits

**Comment by problame** [comment]
What version of ZFS / FreeBSD are you on?
There is no `dump_bytes` in 13, so I guess it's 12.2?
But

https://github.com/freebsd/freebsd-src/blob/2430d82b40ea1228b4b3dcb88388693262461eb1/sys/cddl/contrib/opensolaris/uts/common/fs/zfs/dmu_send.c#L114

doesn't have a call to the taskq APIs AFAICT... :thinking:

**Comment by problame** [comment]
Ah, gotcha, it's probably FreeBSD 13: `dmu_send_outparams_t::dso_outfunc` => https://github.com/openzfs/zfs/blob/a631283b745efc8354ad680b3a3c1f7bff38c4f8/module/zfs/zfs_ioctl.c#L5354

**Comment by problame** [comment]
Curious, so the OpenZFS 'redacted send and recv commit' https://github.com/openzfs/zfs/commit/30af21b02569ac192f52ce6e6511015f8a8d5729 introduced the behavior where.

Both Linux and FreeBSD use `spa_taskq_dispatch_sync` => `taskq_wait_id`.
On Linux, `taskq_wait_id` uses `wait_event`, which is uninterruptible.
And as @avg-I pointed out, the FreeBSD impl of `taskq_wait_id` is also waiting uninterruptible.
Hence, `dump_bytes_cb` ([link](https://github.com/openzfs/zfs/blob/a631283b745efc8354ad680b3a3c1f7bff38c4f8/module/zfs/zfs_ioctl.c#L5323-L5324)) needs to always return or else `zfs send` is stuck in uninterruptible state.

Unless Linux is doing some funky synergy between SIGKILL and pipe writes, I suspect both platforms are affected by this.

**Comment by problame** [comment]
Hah, of course Linux does the check, and the write is interruptible:

https://github.com/torvalds/linux/blob/5b697f86f9f136d200c9827d6eca0437b7eb96cf/fs/pipe.c#L543-L547

https://github.com/torvalds/linux/blob/5b697f86f9f136d200c9827d6eca0437b7eb96cf/fs/pipe.c#L560

**Comment by problame** [comment]
Hm, so are pipe writes on FreeBSD always blocking? I would doubt it...

**Comment by problame** [comment]
WIP of my commit message:


```
    On FreeBSD with OpenZFS 2.0, we might wait indefinitely for `zfs send` to exit.
    The reason is that, due to the refactoring done for redacted send & recv
    (https://github.com/openzfs/zfs/commit/30af21b02569ac192f52ce6e6511015f8a8d5729),
    the `dump_bytes` function, which writes to the pipe, executes in a separate thread (synctask taskq).
    The `zfs send` process/thread waits for that taskq thread using an uninterruptible primitive.
    So when we SIGKILL `zfs send`, that signal doesn't reach the right thread to interrupt the pipe write.
    On Linux, that doesn't appear to be a problem and `zfs send` reacts to SIGKILL (why?).
    But on FreeBSD, `zfs send` simply cannot be killed.
```

**Comment by problame** [comment]
We can `SIGKILL` on Linux because, `ifdef HAVE_LARGE_STACKS`, which is true at least on my system, the `zfs send` thread performs the `pipe_write` and the taskq is not used.

https://github.com/openzfs/zfs/blob/a631283b745efc8354ad680b3a3c1f7bff38c4f8/module/zfs/zfs_ioctl.c#L5345-L5347

```
cs@cstp:[~/zrepl/zrepl]: sudo cat /proc/$(pgrep -f 'zfs send.*platform.*')/stack
[sudo] password for cs: 
[<0>] pipe_write+0x2af/0x560
[<0>] new_sync_write+0x105/0x180
[<0>] __kernel_write+0xce/0x160
[<0>] kernel_write+0x74/0x100
[<0>] vn_rdwr+0x48/0xa0
[<0>] dump_bytes_cb+0x63/0xc0
[<0>] dump_record+0xfb/0x160
[<0>] dump_write+0xde/0x340
[<0>] do_dump+0x6eb/0xa30
[<0>] dmu_send_impl+0x5ca/0xb20
[<0>] dmu_send_obj+0x1c9/0x2e0
[<0>] zfs_ioc_send+0x9c/0x290
[<0>] zfsdev_ioctl+0x7ca/0x8e0
[<0>] __x64_sys_ioctl+0x83/0xb0
[<0>] do_syscall_64+0x33/0x40
[<0>] entry_SYSCALL_64_after_hwframe+0x44/0xa9
```

**Comment by problame** [comment]
I opened upstream bug https://github.com/openzfs/zfs/issues/12500

It turns out that, [on July 16, the FreeBSD-specific bits of OpenZFS got `HAVE_LARGE_STACKS`](https://github.com/openzfs/zfs/pull/12350/files).
That means that the bug will magically go away as soon as you upgrade.

I'll be dabbling around with a zrepl workaround though, since AFAIK, 13-RELEASE won't get significant ZFS updates (?).

**Comment by problame** [comment]
@avg-l I opened a PR with a fix & test coverage.
Please run the zrepl platformtests, and then zrepl itself, on your deployment.
You can use the artifacts produced by CI or build your own.

**Comment by avg-I** [comment]
That was an interesting investigation and a lot of details.
I am testing 495-fix-stuck-waiting-for-zfs-send-on-freebsd.
Thank you!

**Comment by avg-I** [comment]
And, just in case, I saw the problem on FreeBSD 14.0-CURRENT arm64 from early June.

**Comment by problame** [comment]
We really need to get CI for the platformtests going.
If you can do the testing on FreeBSD 14 as well, that'd be awesome.

**Comment by problame** [comment]
@avg-I updates on this?

**Comment by barrylustig** [comment]
> @avg-I updates on this?

FWIW, I've been running with the PR you put up 3 weeks ago on my FreeBSD 13 system.  It's been rock solid since then.  Before, I could count on zrepl locking up with 2-3 days.

**Comment by problame** [comment]
@barrylustig pleased to hear that. Are you running it on multiple cores or are you using `cpuset` to work around #411 ?

**Comment by barrylustig** [comment]
> @barrylustig pleased to hear that. Are you running it on multiple cores or are you using `cpuset` to work around #411 ?

Multiple cores.

**Comment by avg-I** [comment]
Thank you for the fix and sorry for a quite late feedback.
I tested the latest zrepl by placing a firewall between the backup server and a test machine.
I configured bandwidth throttling and random packet drop to simulate various degrees of network flakiness.

I saw lots of different zrepl errors in the test. As far I can tell, all of them were correctly handled.
zrepl never hanged and either managed to "push through" or gave up on a current attempt.

**Comment by avg-I** [comment]
> We really need to get CI for the platformtests going. If you can do the testing on FreeBSD 14 as well, that'd be awesome.

@problame, when you said that, did you mean the test for the problem described in this issue or did you mean tests run by test-platform-bin  make target?


P.S.
FWIW, I am able to build zrepl without any issues for GOOS=freebsd GOARCH=arm GOARM=7 and GOOS=freebsd   GOARCH=arm64. And they even work for me :-)
So, you can add them to release targets if you'd like.
Those builds are actually cross-builds on a FreeBSD 13.0 amd64 host. Not sure if that matters.

**Comment by problame** [comment]
Glad to hear that the fix works for you.

With "running platformtests" I mean the following:

```
make test-platform-bin && sudo make test-platform 
```

**Comment by problame** [comment]
>  So, you can add them to release targets if you'd like.

Will create a separate issue for that.
